File: crates/python/mirascript/_vm/lib/vm_global/sequence/group.py
from __future__ import annotations

from ....._helpers.convert import to_string
from ....._helpers.constants import Uninitialized
from ....operations import Call
from ....types import VmAny, VmConst
from ..._helpers import _expect_array, _expect_callable


def group_by(data: VmAny = Uninitialized, key_fn: VmAny = Uninitialized):
    data = _expect_array("data", data, None)
    key_fn = _expect_callable("key_fn", key_fn, data)

    result: dict[str, list[VmConst]] = {}

    for i, item in enumerate(data):
        if item is Uninitialized:
            item = None
        key = to_string(Call(key_fn, item, i, data))
        l = result.setdefault(key, [])
        l.append(item)

    return result

File: packages/mirascript/src/vm/lib/global/sequence/group.ts
import { toString } from '../../../../helpers/convert/index.js';
import { fromEntries } from '../../../../helpers/utils.js';
import { $Call } from '../../../operations/call.js';
import type { VmConst } from '../../../types/index.js';
import { VmLib, expectArray, expectCallable } from '../../helpers.js';

export const group_by = VmLib(
    (data, key) => {
        expectArray('data', data, null);
        expectCallable('key', key, data);
        const result = new Map<string, VmConst[]>();
        const len = data.length;
        for (let i = 0; i < len; i++) {
            const v = data[i] ?? null;
            const k = toString($Call(key, [v, i, data]), undefined);
            let arr = result.get(k);
            if (!arr) {
                arr = [];
                result.set(k, arr);
            }
            arr.push(v);
        }
        return fromEntries(result);
    },
    {
        summary: '根据指定的键函数对数组进行分组',
        params: {
            data: { type: 'array', description: '要分组的数组' },
            key: {
                type: 'fn(value: any, index: number, arr: type(data)) -> string',
                description: '用于生成分组键的函数',
            },
        },
        returns: { type: 'record' },
        examples: ['group_by([1, 2, 3, 4], fn { it % 2 }) // ("0": [2, 4], "1": [1, 3])'],
    },
);
